File: mlir/examples/dsp/SimpleBlocks/Output/TryDSPApps/BenchmarkTest/CCode/digitalModulation.c
#include <stdio.h>
#include <stdlib.h>
#include <complex.h>
#include <time.h>

#define INPUT_LENGTH 100000000

void generate_input(int *data) {
    for(int i=0; i<INPUT_LENGTH; ++i) {
        data[i] = rand() %2;
    }
}

void qam_modulate(complex double *symbols, int *data) {
    for(int i=0; i<INPUT_LENGTH; i+=2) {
        int bit1 = data[i];
        int bit2 = data[i+1];

        if(bit1 == 0 && bit2==0)
            symbols[i/2] = -1 - 1*I;
        else if(bit1 == 0&& bit2==1)
            symbols[i/2] = -1 + 1*I;
        else if(bit1==1 && bit2==0)
            symbols[i/2] = 1-1*I;
        else if (bit1==1 && bit2==1)
            symbols[i/2] = 1+1*I;
    }
}

void qam_demodulate(int *bits, complex double *symbols) {
    for (int i = 0; i < INPUT_LENGTH/2; i++) {
        complex double symbol = symbols[i];

        if (symbol == -1 - 1*I) {
            bits[2 * i] = 0;
            bits[2 * i + 1] = 0;
        } else if (symbol == -1 + 1*I) {
            bits[2 * i] = 0;
            bits[2 * i + 1] = 1;
        } else if (symbol == 1 - 1*I) {
            bits[2 * i] = 1;
            bits[2 * i + 1] = 0;
        } else if (symbol == 1 + 1*I) {
            bits[2 * i] = 1;
            bits[2 * i + 1] = 1;
        }
    }
}

int main() {
    srand(time(NULL)); // Seed random number generator
    
    int* data;
    data = (int*)malloc(sizeof(int) * INPUT_LENGTH);
    complex double* symbols;
    symbols = (complex double*)malloc(sizeof(complex double) * INPUT_LENGTH / 2);
    
    // Generate random input data
    generate_input(data);
    
    // Perform QAM modulation
    qam_modulate(symbols, data);
    
    // Perform QAM demodulation
    int* bits; 
    bits = (int*)malloc(sizeof(int) * INPUT_LENGTH);
    qam_demodulate(bits, symbols);


    printf("%d ", bits[5]);
    
    
    
    
    return 0;
}

File: mlir/examples/dsp/SimpleBlocks/Output/TryDSPApps/BenchmarkTest/Matlab/digitalModulation.m
clc;
clear;

% Define constants
INPUT_LENGTH = 100000000;

% Generate random input data
data = randi([0 1], 1, INPUT_LENGTH);

% QAM Modulation
function symbols = qam_modulate(data)
    symbols = zeros(1, length(data)/2);
    for i = 1:2:length(data)
        bit1 = data(i);
        bit2 = data(i+1);
        
        if bit1 == 0 && bit2 == 0
            symbols((i+1)/2) = -1 - 1i;
        elseif bit1 == 0 && bit2 == 1
            symbols((i+1)/2) = -1 + 1i;
        elseif bit1 == 1 && bit2 == 0
            symbols((i+1)/2) = 1 - 1i;
        elseif bit1 == 1 && bit2 == 1
            symbols((i+1)/2) = 1 + 1i;
        end
    end
end

% QAM Demodulation
function bits = qam_demodulate(symbols)
    bits = zeros(1, length(symbols)*2);
    for i = 1:length(symbols)
        symbol = symbols(i);
        
        if symbol == -1 - 1i
            bits(2*i-1) = 0;
            bits(2*i) = 0;
        elseif symbol == -1 + 1i
            bits(2*i-1) = 0;
            bits(2*i) = 1;
        elseif symbol == 1 - 1i
            bits(2*i-1) = 1;
            bits(2*i) = 0;
        elseif symbol == 1 + 1i
            bits(2*i-1) = 1;
            bits(2*i) = 1;
        end
    end
end

% Main script
rng('shuffle'); % Seed random number generator

% Perform QAM modulation
symbols = qam_modulate(data);

% Perform QAM demodulation
bits = qam_demodulate(symbols);

% Print the 6th bit (equivalent to bits[5] in C, as MATLAB uses 1-based indexing)
disp(bits(6));